ui/src/app/views/notebook-resource-card.tsx

```diff
@@ -0,0 +1,188 @@
+      const workspacePrefix = `/workspaces/${workspaceNamespace}/${workspaceFirecloudName}`;
+
+      const queryParams = new URLSearchParams([
+        ['playgroundMode', 'false'],
+        ['jupyterLabMode', String(jupyterLab)]
+      ]);
+
+      if (this.readOnly) {
+        queryParams.set('jupyterLabMode', 'true');
+      }
+      return `${workspacePrefix}/notebooks/preview/${encodeURIComponent(notebook.name)}`;
+    }
+
+    async receiveNotebookRename(newName) {
+      const {resourceCard} = this.props;
+      try {
+        await workspacesApi().renameNotebook(
+          resourceCard.workspaceNamespace,
+          resourceCard.workspaceFirecloudName,
+          {
+            name: resourceCard.notebook.name,
+            newName: this.fullNotebookName(newName)
+          });
+      } catch (error) {
+        console.error(error); // TODO: better error handling
+      } finally {
+        this.setState({renaming: false});
+        this.props.onUpdate();
+      }
+    }
+
+    async deleteNotebook(onComplete: Function) {
+      workspacesApi().deleteNotebook(
+        this.props.resourceCard.workspaceNamespace,
+        this.props.resourceCard.workspaceFirecloudName,
+        this.props.resourceCard.notebook.name)
+        .then(() => {
+          onComplete();
+          this.props.onUpdate();
+        });
+    }
+
+    async duplicateNotebook() {
+      workspacesApi().cloneNotebook(
+        this.props.resourceCard.workspaceNamespace,
+        this.props.resourceCard.workspaceFirecloudName,
+        this.props.resourceCard.notebook.name)
+        .then(() => {
+          this.props.onUpdate();
+        }).catch(e => {
+          this.props.onDuplicateResource(false);
+          this.props.showErrorModal('Duplicating Notebook Error',
+            'Notebook with the same name already exists.');
+        });
+    }
+
+    fullNotebookName(name) {
+      return !name || /^.+\.ipynb$/.test(name) ? name : `${name}.ipynb`;
+    }
+
+    render() {
+      return <React.Fragment>
+        {this.state.showCopyNotebookModal &&
+        <CopyNotebookModal
+          fromWorkspaceNamespace={this.props.resourceCard.workspaceNamespace}
+          fromWorkspaceName={this.props.resourceCard.workspaceFirecloudName}
+          fromNotebook={this.props.resourceCard.notebook}
+          onClose={() => this.setState({showCopyNotebookModal: false})}
+          onCopy={() => this.props.onUpdate()}/>
+        }
+
+        {this.state.renaming &&
+        <RenameModal type='Notebook'
+                     onRename={(newName) => this.receiveNotebookRename(newName)}
+                     onCancel={() => this.setState({renaming: false})}
+                     hideDescription={true}
+                     oldName={this.props.resourceCard.notebook.name}
+                     nameFormat={(name) => this.fullNotebookName(name)}/>
+        }
+
+        <ResourceCardTemplate
+          actions={this.actions}
+          actionsDisabled={!this.writePermission}
```

Comment:
I wonder if it might be more flexible for each action to have its own logic for deciding if it should be enabled or not. There are some scenarios where only a subset of actions should be enabled, right?

Reply:
That would be cool. I'll add that refactoring when I run into those actions.